### PR TITLE
Add new rate limiting strategies: Fixed Window and Token Bucket for both in-memory and Redis stores

### PR DESCRIPTION
**Pull request by seifalmotaz**
This pull request introduces two new rate limiting strategies to the Robyn Rate Limits extension: Fixed Window and Token Bucket. These strategies are implemented for both in-memory and Redis-based stores, providing users with more flexibility in managing their API rate limits.

Key changes:
1. Added InMemoryFixedWindowStore and InMemoryTokenBucketStore classes for in-memory rate limiting.
2. Added RedisFixedWindowStore and RedisTokenBucketStore classes for Redis-based rate limiting.
3. Updated the existing sliding window implementations (InMemoryStore and RedisStore) for consistency and improved performance.
4. Updated the README.md to include documentation and usage examples for all new rate limiting strategies.

These additions allow users to choose the most appropriate rate limiting strategy for their specific use case, whether they need simple fixed windows, allow for traffic bursts with token buckets, or prefer the accuracy of sliding windows.

All new implementations have been tested and are compatible with the existing RateLimiter class. The Redis-based implementations use pipelines and transactions for improved efficiency and thread safety.

### CODE REVIEW

File: robyn_rate_limits/in_memory.py [hunk 1]
```diff
@@ -1,14 +1,92 @@
+from typing import List
+from typing import NamedTuple
+from typing import Optional
+
 import cachetools
 
+from robyn_rate_limits.protocols import LimitStore
+
+
+class InMemoryStore(LimitStore):
```
Comment (IdoKendo) [inline robyn_rate_limits/in_memory.py:10]: When working with protocols, you don't need to inherit from the base class, adhering to the protocol is implied by shape.
This is true for all of the rest of the implementations, I will not litter the review with this comment, but it's true to all of them.

File: robyn_rate_limits/protocols.py [hunk 2]
```diff
@@ -2,5 +2,8 @@
 
 
 class LimitStore(Protocol):
-    def get_calls_count(self, limit_key: str, current_timestamp: int) -> int:
```
Comment (IdoKendo) [inline robyn_rate_limits/protocols.py:5]: Can you please revert styling changes? I see many places where the change is adding a trailing command or removing it which changes the way `ruff` formats the files. Please let it determine the line length by itself and don't change it by manipulating trailing commas.

**Review by IdoKendo — CHANGES_REQUESTED** [review]
Hey @seifalmotaz 👋 Amazing work and a huge thank you for contributing!
I had one issue regarding protocol inheritance and a request regarding styling that needs to be fixed, but otherwise it looks really good and I really appreciate the effort!
Let me know if anything's unclear!

**Comment by seifalmotaz** [comment]
Apologies for the delayed response. Could you please review the changes and let me know if there's anything else you'd like to adjust?

**Comment by IdoKendo** [comment]
Hey @seifalmotaz 👋 Thanks again for the contribution! It looks good on my part - just please fix the linting error in the pipeline and I will merge.

**Review by IdoKendo — COMMENTED** [review]
Hey @seifalmotaz !
Are you still planning on following through with this PR?

**Comment by seifalmotaz** [comment]
> Hey @seifalmotaz !
> Are you still planning on following through with this PR?

Yes, i am. Sorry for being. I have been busy.

Could you explain what the problem is with linting because i don't fully understand the problem?

**Comment by IdoKendo** [comment]
> > Hey @seifalmotaz !
> > Are you still planning on following through with this PR?
> 
> Yes, i am. Sorry for being late. I have been busy.
> 
> Could you explain what the problem is with linting because i don't fully understand the problem?

You can see in the pipeline:
```
 lint: commands[0]> black --check robyn_rate_limits/
  would reformat /Users/runner/work/robyn_rate_limits/robyn_rate_limits/robyn_rate_limits/in_memory.py
  
  Oh no! 💥 💔 💥
```

You can run black locally to fix it. Also, I've migrated away from black to ruff (the formatting rules are aligned) so if you rebase on main you can run ruff format instead.
Let me know if you need further help!

**Pull request merged**